backends/oauth2.go

```diff
@@ -0,0 +1,361 @@
+package backends
+
+import (
+	"context"
+	"encoding/json"
+	. "github.com/iegomez/mosquitto-go-auth/backends/constants"
+	"github.com/iegomez/mosquitto-go-auth/backends/topics"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	go_oauth2 "golang.org/x/oauth2"
+	go_clientcredentials "golang.org/x/oauth2/clientcredentials"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type userState struct {
+	username          string
+	superuser         bool
+	readTopics        []string
+	writeTopics       []string
+	subscribeTopics   []string
+	denyTopics        []string
+	lastUserInfoUpate time.Time
+	createdAt         time.Time
+	updatedAt         time.Time
+	client            *http.Client
+	token             *go_oauth2.Token
+}
+
+type UserInfo struct {
+	Sub  string `json:"sub"`
+	MQTT struct {
+		Topics struct {
+			Read      []string `json:"read"`
+			Write     []string `json:"write"`
+			Subscribe []string `json:"subscribe"`
+			Deny      []string `json:"deny"`
+		} `json:"topics"`
+		Superuser bool `json:"superuser"`
+	} `json:"mqtt"`
+}
+
+type Oauth2 struct {
+	tokenUrl      string
+	userInfoURL   string
+	userCache     map[string]userState
+	cacheDuration time.Duration
+	version       string
+	scopesSplit   []string
+}
+
+func NewOauth2(authOpts map[string]string, logLevel log.Level) (*Oauth2, error) {
+	log.SetLevel(logLevel)
+
+	var oauth2 = &Oauth2{}
+	oauth2.version = "1.0.0"
+
+	missingOpts := ""
+	oauth2Ok := true
+
+	tokenUrl, ok := authOpts["oauth_token_url"]
+	if ok {
+		oauth2.tokenUrl = tokenUrl
+	} else {
+		oauth2Ok = false
+		missingOpts += " oauth_token_url"
+	}
+
+	if userInfoURL, ok := authOpts["oauth_userinfo_url"]; ok {
+		oauth2.userInfoURL = userInfoURL
+	} else {
+		oauth2Ok = false
+		missingOpts += " oauth_userinfo_url"
+	}
+
+	if cacheDurationSeconds, ok := authOpts["oauth_cache_duration"]; ok {
+		if durationInt, err := strconv.Atoi(cacheDurationSeconds); err == nil {
+			oauth2.cacheDuration = time.Duration(durationInt) * time.Second
+		} else {
+			log.Errorf("unable to parse cacheDurationSeconds: %s", err)
+		}
+	} else {
+		oauth2Ok = false
+		missingOpts += " oauth_cache_duration"
+	}
+
+	if scopes, ok := authOpts["oauth_scopes"]; ok {
+		oauth2.scopesSplit = strings.Split(strings.Replace(scopes, " ", "", -1), ",")
+	} else {
+		log.Infof("no scopes where specified")
+		oauth2.scopesSplit = []string{}
+	}
+
+	oauth2.userCache = make(map[string]userState)
+
+	if !oauth2Ok {
+		return oauth2, errors.Errorf("Oauth2 backend error: missing remote options: %s", missingOpts)
+	}
+
+	return oauth2, nil
+}
+
+func (o *Oauth2) GetUser(username, password, clientid string) (bool, error) {
+	if password == "oauthbearer_empty_password" {
+		return o.createUserWithToken(username, clientid)
+	} else {
+		return o.createUserWithClientCredentials(username, password, clientid)
+	}
+}
+
+func (o *Oauth2) GetSuperuser(username string) (bool, error) {
+	// Function that checks if the user has admin privilies
```

Comment:
The convention in Go is to annotate exported functions with a comment over the signature so the tooling may automatically convert them into documentation. Many times it's rather useless since the function/method name is clear enough, so we just do it to keep the formatter/linter from yelling. Having the comment inside the function is not really informative and should probably be removed.